### ISSUE TITLE
TCP connection closed errors not surfaced to the consumer

### ISSUE THREAD

**Issue by minhajuddin** [rust-lang/rust#93992]
## Summary
When a TCP connection is closed (`FIN,ACK`) from the consumer it is `ACK`ed by the rust server, but writing to this stream after this does not raise an error.

Here is a sequence of events:

1. Rust TCP server starts listening to port 4050.
2. A connection is made from netcat to this port which is successful.
3. netcat sends a payload which is successfully read by the server. At this point the server sleeps for 5 seconds (this is to give us time to close netcat)
4. We close netcat which sends a `FIN,ACK` to the server. The server responds with an `ACK`. At this point the connection should be closed.
5. After sleeping for 5 seconds, the connection handler resumes and writes to this tcp stream, *this should raise an error* because the connection has been closed (the server has even returned an ACK). However, at this point no errors are returned. On the TCP layer, a packet is pushed to the consumer and the OS returns a RST (because netcat is dead)

I have uploaded 2 screenshots. The first which shows the execution of this program.
![image](https://user-images.githubusercontent.com/43916/153892224-1bc38a66-1a79-4770-a2de-7152506ca31e.png)

Here is the image containing the TCP traffic from wireshark.
![image](https://user-images.githubusercontent.com/43916/153892291-deac4019-b7fc-4a50-97ae-ff82a6038fcc.png)

Here is the code for the example client handler. The whole repo can be found at https://github.com/minhajuddin/tcpbug/blob/main/src/main.rs

```rust
fn handle_client(mut stream: TcpStream) -> Result<(), Box<dyn Error>> {
    let mut buffer = [0; 512];
    let byte_count = stream.read(&mut buffer)?;

    println!("Received {} bytes", byte_count);

    let response = String::from_utf8_lossy(&buffer[..]);
    println!("Received: {}", response);

    println!("Sleeping for 5 seconds, Close netcat on the other end");
    thread::sleep(time::Duration::from_secs(5));

    println!("Will try to send response. THIS SHOULD FAIL!");
    stream.write_all(response.as_bytes())?;
    stream.flush()?;
    println!("Expected errrors were not returned!");

    Ok(())
}

```

### Rust version info
```
$ rustc --version --verbose
rustc 1.58.1 (db9d1b20b 2022-01-20)
binary: rustc
commit-hash: db9d1b20bba1968c1ec1fc49616d4742c1725b4b
commit-date: 2022-01-20
host: x86_64-unknown-linux-gnu
release: 1.58.1
LLVM version: 13.0.0
```

**Comment by hkratz** [comment]
This behaves as expected as far as I can see. The client has requested close with FIN which the server has acknowledged with ACK. The client has received this ACK. Now the client is in the FIN-WAIT-2 state and the server is in the CLOSE-WAIT state. In those states the server can still send data ([TCP spec - SEND Call](https://datatracker.ietf.org/doc/html/rfc793#page-56)) and the client can still receive it ([TCP spec - RECEIVE Call](https://datatracker.ietf.org/doc/html/rfc793#page-58)).

Only once the server has sent FIN itself (by closing the sending part on the server side, in Rust via `stream.shutdown(Shutdown::Write)` or `stream.shutdown(Shutdown::Both)`) it can no longer send data back to the client.

**Comment by Enselic** [comment]
Triage: Let's close as "works as designed" then.
